src/compiler/types.ts

```diff
@@ -1176,7 +1184,7 @@ namespace ts {
 
     export interface StringLiteral extends LiteralExpression {
         kind: SyntaxKind.StringLiteral;
-        /* @internal */ textSourceNode?: Identifier | StringLiteralLike | NumericLiteral; // Allows a StringLiteral to get its text from another node (used by transforms).
+        /* @internal */ textSourceNode?: Identifier | PrivateName | StringLiteralLike | NumericLiteral; // Allows a StringLiteral to get its text from another node (used by transforms).

```

Comment:
I don't think this case should occur (?)

Reply:
@RyanCavanaugh thanks for noticing this, I wasn't sure what to do here. In factory.ts there is a helper function for creating literals from nodes, it seemed that we would want this to work from private names as well:

https://github.com/bloomberg/TypeScript/pull/4/files#diff-9c914fcc05c72a72db62c0a023b15f3fR110

Reply:
ping @RyanCavanaugh

Reply:
`createLiteralFromNode` shouldn't accept a `PrivateName` node - unlike property names (whose names are meaningful in isolation), private names can't be used anywhere a literal is needed.

Any place where `createLiteral` is being called today with a node that is sourced from a property name, we need to add logic to do the right thing for `PrivateName`s. This should just be mostly `assert`s to make sure we don't e.g. try to rewrite `x.#foo` with `x["#foo"]` (something we currently treat as being always OK).

Reply:
I incorporated this feedback and [disallowed creating literals from private names](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff).

I think the code does what we want now, with these caveats:
- [this type alias smells a little](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff#diff-9c914fcc05c72a72db62c0a023b15f3fR3585). I used it to avoid making big changes to the type hierarchy in types.d.ts.
- I also had to stick a [TODO in the transformer](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff#diff-db4815b580d3d6fd21c975e36c8f81d1R1645).

Reply:
ping @RyanCavanaugh would you be able to take another look?